src/core/lib/surface/call.c

```diff
@@ -1384,12 +1383,28 @@ static void receiving_initial_metadata_ready(grpc_exec_ctx *exec_ctx,
     }
   }
 
-  call->has_initial_md_been_received = true;
-  if (call->saved_receiving_stream_ready_bctlp != NULL) {
-    grpc_closure *saved_rsr_closure = GRPC_CLOSURE_CREATE(
-        receiving_stream_ready, call->saved_receiving_stream_ready_bctlp,
-        grpc_schedule_on_exec_ctx);
-    call->saved_receiving_stream_ready_bctlp = NULL;
+  grpc_closure *saved_rsr_closure = NULL;
+  while (true) {
```

Comment:
Would this code be simpler if we kept `saved_receiving_stream_ready_bctlp` as a separate non-atomic pointer, and used the atomic value only for an enum indicating the state?  I'm thinking of an enum like this:

```
typedef enum {
  RECV_NONE,
  RECV_INITIAL_METADATA_FIRST,
  RECV_MESSAGE_FIRST,
} recv_state;
```

Then the fields of `grpc_call` are:

```
gpr_atm recv_state;  // a recv_state enum
void *saved_receiving_stream_ready_bctlp;
```

Then the code in `receiving_stream_ready()` becomes:

```
call->saved_receiving_stream_ready_bctlp = bctlp;
if (error != GRPC_ERROR_NONE || call->receiving_stream == NULL ||
    !gpr_atm_full_cas(&call->recv_state, RECV_NONE, RECV_MESSAGE_FIRST)) {
  call->saved_receiving_stream_ready_bctlp = NULL;
  process_data_after_md(exec_ctx, bctlp);
}
```

And the code here becomes:

```
if (!gpr_atm_full_cas(&call->recv_state, RECV_NONE, RECV_INITIAL_METADATA_FIRST)) {
  grpc_closure *saved_rsr_closure = GRPC_CLOSURE_CREATE(
      receiving_stream_ready, call->saved_receiving_stream_ready_bctlp,
      grpc_schedule_on_exec_ctx);
  call->saved_receiving_stream_ready_bctlp = NULL;
  GRPC_CLOSURE_RUN(exec_ctx, saved_rsr_closure, GRPC_ERROR_REF(error));
}
```

What do you think?